bdai_ros2_wrappers/bdai_ros2_wrappers/executors.py

```diff
@@ -0,0 +1,610 @@
+        Returns:
+            A future for the result of the work submitted.
+
+        Raises:
+            RuntimeError: if the pool has been shutdown.
+        """
+        with self._submit_lock, self._shutdown_lock:
+            if self._shutdown:
+                raise RuntimeError("cannot submit to a shutdown pool")
+            future: concurrent.futures.Future = concurrent.futures.Future()
+            work = AutoScalingThreadPool.Work(future, fn, args, kwargs)
+            self._logger.debug(
+                f"Submitting work '{work}'...",
+            )
+            if self._submission_quota > len(self._runlists[work.fn]):
+                if work.fn in self._waitqueues and self._waitqueues[work.fn]:  # prioritize pending work
+                    self._logger.debug("Have similar work pending")
+                    self._logger.debug(f"Work '{work}' put to wait", work)
+                    self._waitqueues[work.fn].append(work)
+                    self._logger.debug("Fetching pending work...")
+                    work = self._waitqueues[work.fn].popleft()
+                    while work.cancelled() and self._waitqueues[work.fn]:
+                        self._logger.debug(f"Work '{work}' was cancelled, notify it")
+                        work.notify_cancelation()
+                        work = self._waitqueues[work.fn].popleft()
+                    if not self._waitqueues[work.fn]:
+                        del self._waitqueues[work.fn]
+                    self._logger.debug(f"Got work '{work}'")
+                if not work.cancelled():
+                    self._logger.debug(f"Proceed with work '{work}'")
+                    self._runlists[work.fn].append(work)
+                    self._do_submit(work)  # actually submit work
+                else:
+                    self._logger.debug(f"Work '{work}' was cancelled, notify it")
+                    work.notify_cancelation()
+            else:
+                self._logger.debug(f"Hit quota for work '{work}'")
+                self._waitqueues[work.fn].append(work)
+                self._logger.debug(f"Work '{work}' put to wait")
+            self._logger.debug(f"Done submitting work '{work}'")
+            return future
+
+    def shutdown(self, wait: bool = True, *, cancel_futures: bool = False) -> None:
+        """
+        Shuts down the pool.
+
+        Args:
+            wait: whether to wait for all worker threads to shutdown.
+            cancel_futures: whether to cancel all ongoing work (and associated futures).
+        """
+        self._logger.debug("Shutting down pool...")
+        with self._shutdown_lock:
+            self._shutdown = True
+        self._logger.debug("Pool shutdown")
+
+        if cancel_futures:
+            with self._submit_lock:
+                self._logger.debug("Canceling all work in progress...")
+                for runlist in self._runlists.values():
+                    for work in runlist:
+                        work.cancel()
```

Comment:
Fortunately, [it is an idempotent method](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.cancel). Checking wouldn't work anyways, as it would be subject to TOCTOU races.